### PR TITLE
proxy protocol probe mvp

### PR DESCRIPTION
**Pull request by AngeloDanducci**
# Changes

- Adds proxy protocol support when probing

/kind enhancement

Fixes
https://github.com/knative/serving/issues/13123
https://github.com/knative-sandbox/net-istio/issues/463

This is my first contribution to networking directly and I have a fair amount of questions, so I'll try and add tasks for questions that I think should be answered before this gets merged.

**New Dependency**

I've added a new [third party dependency](https://github.com/pires/go-proxyproto) that handles the proxy protocol header construction.

I've added the license file to the third party folder, the project is licensed under Apache 2.0 so I assume no issues here.

**Initial testing**

Initial testing scenario is as follows:

1. Knative serving installed normally via yaml
2. net-istio picking up this set of changes locally, use ko to apply the new net-istio controller to my cluster
3. Configure Magic DNS / sslip
4. [apply the proxy protocol envoy filter as seen in the istio documentation](https://istio.io/latest/docs/ops/configuration/traffic-management/network-topologies/#proxy-protocol)
5. create a new test service
6. observe that the test service becomes ready, and responds to normal network traffic

**Design Discussion**

Refactored this to not support the offer/doAsync portion - they don't need to be supported as it's only used by serving's PA and that probes via private service.

Let me know if there's a better way to consume the `config-network` values.
Needed to add a volume mount to support the current approach https://github.com/knative-sandbox/net-istio/pull/1028

Will need to test with Contour but I imagine it's the same as Istio - just need to consume the configmap.

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -295,6 +315,8 @@ func NewConfigFromMap(data map[string]string) (*Config, error) {
 		cm.AsBool(EnableMeshPodAddressabilityKey, &nc.EnableMeshPodAddressability),
 		cm.AsString(DefaultExternalSchemeKey, &nc.DefaultExternalScheme),
 		cm.AsBool(InternalEncryptionKey, &nc.InternalEncryption),
+		cm.AsBool(ProxyProtocolProbeEnabled, &nc.ProxyProtocolProbeEnabled),
+		cm.AsString(ProxyProtocolFilter, &nc.ProxyProtocolFilter),
```
Comment (psschwei) [inline pkg/config/config.go:319]:
```suggestion
		cm.AsBool(ProxyProtocolProbeEnabledKey, &nc.ProxyProtocolProbeEnabled),
		cm.AsString(ProxyProtocolFilterKey, &nc.ProxyProtocolFilter),
```

**Comment by dprotaso** [comment]
Does enabling the proxy protocol break intra-cluster traffic? ie. Knative Service => Knative Service on the same cluster?

**Comment by AngeloDanducci** [comment]
> Does enabling the proxy protocol break intra-cluster traffic? ie. Knative Service => Knative Service on the same cluster?
> 
> Given that the prober fails without this change I'm assuming yes?

Hmm good question.  The default case is nothing changes.
When proxy protocol is enabled, Knative Service => Knative Service traffic should not break if the traffic is going through the service mesh, but could be problematic if services are making raw http requests to one another.

However that should be the same case if it's enabled for any net-* package and we already support proxy protocol with kourier (https://github.com/knative-sandbox/net-kourier/pull/744) and Istio supports enablement through an [envoy filter](https://istio.io/latest/docs/ops/configuration/traffic-management/network-topologies/#proxy-protocol) but net-istio needs the probing change.

Do Knative services probe each other?  If not I don't think this should have any effect on ksvc => ksvc traffic.  This change is limited to probes and how they are processing work items (Is the pod reachable from the prober (ie net-istio controller)?).

File: go.mod [hunk 2]
```diff
@@ -7,6 +7,7 @@ require (
 	github.com/google/go-cmp v0.5.8
 	github.com/gorilla/websocket v1.4.2
 	github.com/hashicorp/golang-lru v0.5.4
+	github.com/pires/go-proxyproto v0.6.2
```
[the remark this answers is not in the document]
Comment (AngeloDanducci) [reply go.mod:10]: Good to be aware of.  We also depend on it for [kourier's implementation of proxy protocol](https://github.com/knative-sandbox/net-kourier/pull/744/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R11)

**Comment by dprotaso** [comment]
> Knative Service => Knative Service traffic should not break if the traffic is going through the service mesh

I'm referring to the case when there is no service mesh and the traffic goes through the istio local gateway (ie. the envoy proxy to take advantage of traffic splitting etc)

**Comment by AngeloDanducci** [comment]
> Does enabling the proxy protocol break intra-cluster traffic? ie. Knative Service => Knative Service on the same cluster?
> 
> Given that the prober fails without this change I'm assuming yes?

Gave testing this a go as follows:

1. Knative serving installed
2. Istio/Net-istio setup
3. Add the envoy filter for [proxy protocol](https://istio.io/latest/docs/ops/configuration/traffic-management/network-topologies/#proxy-protocol) and enable proxy protocol flag as in this PR
4. Spin up two new knative services via kn service create
5. Exec into one of the knative service pods and try to curl the other knative service

**Result**
Succeeds with normal http request on cluster local:
`curl inhouse-00001.default.svc.cluster.local:80`
`HTTP/1.1 200 OK`

Fails with proxy protocol request on cluster local:
`curl inhouse-00001.default.svc.cluster.local:80 -v --haproxy-protocol`
`400 Bad Request`

However if called using the Magic DNS URL the reverse is true
Fails with normal http request with Magic DNS URL:
`curl http://inhouse.default.<snio>.sslip.io/ -v`
`curl: (56) Recv failure: Connection reset by peer`

Succeeds with proxy protocol and Magic DNS URL:
`curl http://inhouse.default.<snip>.sslip.io/ -v --haproxy-protocol`
` HTTP/1.1 200 OK`

@dprotaso let me know if this passes muster for your question, and if not what test(ing) you would like to see instead.

If it passes I'll make the other small changes here and we should be good to go.

**Comment by dprotaso** [comment]
Hey thanks,

```
curl http://inhouse.default.<snio>.sslip.io/ -v
curl: (56) Recv failure: Connection reset by peer
```

This just confirms that in-cluster traffic is broken when the proxy protocol is enabled. Unless all _workloads_ know to speak that protocol. 

I guess my question is what do you think users/operators should expect?

I'm also wondering if having two gateway instances make more sense. One that speaks proxy protocol and one doesn't.

**Comment by AngeloDanducci** [comment]
Given that a user would need to purposefully turn on proxy protocol (labeled as experimental in the Istio docs) and enable proxy protocol probes I think this is unlikely to impact users who are unprepared for potentially breaking svc -> svc traffic on newly created services.

As for expectations I'm not sure, ideally no in cluster traffic would break but that means any existing knative services would need to either be recreated with proxy protocol or have a two-tiered approach where we can handle both proxy protocol and non proxy protocol at the same time (ie two gateway instances - though I think this will still require this change).

I have a product team willing to test this post holiday vacation, let's revisit when they conclude testing - perhaps have a small discussion during the serving meeting with results / how their environment was configured.

**Pull request closed without merging**